### PR TITLE
feat(releaser): add support for custom krew index repo

### LINKED ISSUE #47: Support custom krew index
Thanks for this bot. We are looking to use this bot with our hosted custom plugin index. 
ref: https://github.com/goreleaser/goreleaser

Is that possible? If not, could that be supported?

### PR DESCRIPTION
**Pull request by Gearheads**
### Summary
This PR is to allow users to provide a custom `krew-index` repository, in case they do not wish to upload to the public `krew-index` repository. 
- https://github.com/kubernetes-sigs/krew-index

The goal is to allow users to provide the custom `krew-index` repository information through two new parameters:
- `upstream_krew_index_repo_owner`
- `upstream_krew_index_repo_name`

The GitHub Action will then determine the `default` branch that is configured for the custom `krew-index` repository, and create a PR against that `default` branch. Just in case the custom `krew-index` repository does not use the `master` branch.

This is very helpful for organizations that are hosting their own internal `krew-index` repositories, and allows them to automate their release cycles.

### Fixes
#47 

Please let me know if there is anything else I might be missing. 😄

### CODE REVIEW

File: pkg/krew/repos.go [hunk 1]
```diff
@@ -9,7 +9,7 @@ const (
 
 //GetKrewIndexRepoName returns the krew-index repo name
 func GetKrewIndexRepoName() string {
-	override := os.Getenv("UPSTREAM_KREW_INDEX_REPO_NAME")
+	override := os.Getenv("INPUT_UPSTREAM_KREW_INDEX_REPO_NAME")
```
Comment (rajatjindal) [inline pkg/krew/repos.go:12]: this code actually runs on server side (IIRC), and do not have access to the GitHub inputs. what we will have to do is to accept this input, and then pass along this info to the backend as part of request.

the backend then needs to consider 3 scenarios:

- The override configured on server side. this should take preference over GitHub action input and defaults, as this helps with testing the server.
- The override configured as part of GitHub action (if no override provided on server side)
- The default values. (if no override provided at all)

File: pkg/krew/repos.go [hunk 2]
```diff
@@ -19,7 +19,7 @@ func GetKrewIndexRepoName() string {
 
 //GetKrewIndexRepoOwner returns the krew-index repo owner
 func GetKrewIndexRepoOwner() string {
-	override := os.Getenv("UPSTREAM_KREW_INDEX_REPO_OWNER")
+	override := os.Getenv("INPUT_UPSTREAM_KREW_INDEX_REPO_OWNER")
```
Comment (rajatjindal) [inline pkg/krew/repos.go:22]: same as above

File: pkg/releaser/git.go [hunk 3]
```diff
@@ -123,23 +129,39 @@ func getPushRefSpec(branchName string) string {
...
+    }
+	return repo, nil
+}
+
 // SubmitPR submits the PR
 func (r *Releaser) submitPR(request *source.ReleaseRequest) (string, error) {
-	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: r.Token})
-	tc := oauth2.NewClient(context.TODO(), ts)
-	client := github.NewClient(tc)
+	client := getGitHubClient(r.Token)
```
Comment (rajatjindal) [inline pkg/releaser/git.go:148]: maybe have a client on Releaser object and just reuse that instead of creating here and in clonerepo function. 

also it seems like could needs indentation.

File: pkg/krew/repos.go (hunk 1 above)
Comment (Gearheads) [reply pkg/krew/repos.go]: hi @rajatjindal, thank you so much for the feedback! I have implemented the changes, and the PR should now be updated.

Please let me know if anything else needs to be changed, or if I am missing anything. 😄

**Comment by rajatjindal** [comment]
> Hi @rajatjindal, apologies to bother you, just checking in to see if you have some time to review the updates I made based on your feedback?
> 
> Thank you again for your time, and effort. 😄

Hi @Gearheads this is still on my mind. Apologies for the delay. I will get this tested/reviewed tomorrow.

File: pkg/cicd/circleci/circleci.go [hunk 4]
```diff
@@ -81,3 +81,23 @@ func (p *Provider) GetTemplateFile() string {
 
 	return filepath.Join(p.GetWorkDirectory(), ".krew.yaml")
 }
+
+// GetKrewIndexRepoName gets upstream_krew_index_repo_name
+func (p *Provider) GetKrewIndexRepoName() string {
+	nameInput := getInputForAction("UPSTREAM_KREW_INDEX_REPO_NAME")
```
Comment (rajatjindal) [inline pkg/cicd/circleci/circleci.go:87]: to be consistent with other places, we should keep it same as the input as defined in action.yml, so kindly change it to upstream_krew_index_repo_name

**Comment by rajatjindal** [comment]
Hi @Gearheads 

Thank you for your patience while we review this through. 

tl;dr;

For custom index to work, you need following changes:

- run your own krew-release-bot server side (I am assuming your plugin repo is private, so my bot account won't have access to it anyways)
- When starting the krew-release-bot on server side, add following overrides that can be configured via env variables:

- GH_TOKEN 
- UPSTREAM_KREW_INDEX_REPO_OWNER
- UPSTREAM_KREW_INDEX_REPO_NAME

AND

`KREW_RELEASE_BOT_WEBHOOK_URL` - on your GitHub actions side (e.g. 

Detailed description:

To clarify a few things, I want to write down my thoughts on how this should work and I would love to hear your thoughts if you think some (or all) of it does not make sense.

krew-release-bot has two components:

1. **GitHub action** - This runs on GitHub actions (or travis or circle) - We need the release details to be able to compile the template and render it as per the new release. This is also responsible for sending the request to the backend server.
2. **Server side** - This runs on server side - This code requires details of the repo where we want to open the PR. In this case it would be custom krew-index repo owner/name and GH_TOKEN that has access to that repo.

For custom index scenario, thinking out loud. Lets say we have plugin at `fidelity/my-awesome-plugin` which has a krew template file. Now when we publish a new release, it will fetch the release details, and render the new template.

As a next step, the actions would now submit this request to the server side code. Now there are a couple of things to consider here:

- The server side code has to be running with the GitHub token that has permissions to submit PR to `fidelity/krew-index` repo. 
- The server side needs to know that user want to submit the request to `fidelity/krew-index` repo

so I think we need following changes for this to work:

- You need to run your own krew-release-bot server side (with proper auth to submit PR for `fidelity/krew-index`). You can use env variable GH_TOKEN when starting your server for this.
- You need to override GitHub actions to send the request to the krew-release-bot server that you are running. You can override using environment variable `KREW_RELEASE_BOT_WEBHOOK_URL` when running the GitHub action. 
- You need to override Krew index repo. You can use environment variables `UPSTREAM_KREW_INDEX_REPO_NAME` and `UPSTREAM_KREW_INDEX_REPO_OWNER` for overriding these environment variables on server side.

**Comment by rajatjindal** [comment]
I am also happy to pair on this and get this working for you. (and in the process we can fix any issues that we might encounter).

also happy to chat async on CNCF (or Kubernetes) slack regarding this. Please feel free to DM me there.

File: pkg/cicd/circleci/circleci.go (hunk 4 above)
Comment (Gearheads) [reply pkg/cicd/circleci/circleci.go:87]: no problem. I will update the fields to be lowercase.
